v2/pkg/protocols/others/websocket/websocket.go

```diff
@@ -0,0 +1,365 @@
+func (r *Request) executeRequestWithPayloads(input, hostname string, dynamicValues, previous output.InternalEvent, callback protocols.OutputEventCallback) error {
+	header := http.Header{}
+
+	payloadValues := make(map[string]interface{})
+	for k, v := range dynamicValues {
+		payloadValues[k] = v
+	}
+	parsed, err := url.Parse(input)
+	if err != nil {
+		return errors.Wrap(err, "could not parse input url")
+	}
+	payloadValues["Hostname"] = parsed.Host
+	payloadValues["Host"] = parsed.Hostname()
+	payloadValues["Scheme"] = parsed.Scheme
+	payloadValues["Path"] = parsed.Path
+
+	for key, value := range r.Headers {
+		finalData, dataErr := expressions.EvaluateByte([]byte(value), payloadValues)
+		if dataErr != nil {
+			r.options.Output.Request(r.options.TemplateID, input, "websocket", dataErr)
+			r.options.Progress.IncrementFailedRequestsBy(1)
+			return errors.Wrap(dataErr, "could not evaluate template expressions")
+		}
+		header.Set(key, string(finalData))
+	}
+	websocketDialer := ws.Dialer{
+		Header:    ws.HandshakeHeaderHTTP(header),
+		Timeout:   time.Duration(r.options.Options.Timeout) * time.Second,
+		NetDial:   r.dialer.Dial,
+		TLSConfig: &tls.Config{InsecureSkipVerify: true, ServerName: hostname},
+	}
+
+	finalAddress, dataErr := expressions.EvaluateByte([]byte(r.Address), payloadValues)
+	if dataErr != nil {
+		r.options.Output.Request(r.options.TemplateID, input, "websocket", dataErr)
+		r.options.Progress.IncrementFailedRequestsBy(1)
+		return errors.Wrap(dataErr, "could not evaluate template expressions")
+	}
+	addressToDial := string(finalAddress)
+	if parsed.Path != "" && parsed.Path != "/" {
+		addressToDial = addressToDial + parsed.Path
+	}
+
+	conn, readBuffer, _, err := websocketDialer.Dial(context.Background(), addressToDial)
+	if err != nil {
+		r.options.Output.Request(r.options.TemplateID, input, "websocket", err)
+		r.options.Progress.IncrementFailedRequestsBy(1)
+		return errors.Wrap(err, "could not connect to server")
+	}
+	defer conn.Close()
+
+	responseBuilder := &strings.Builder{}
+	if readBuffer != nil {
+		_, _ = io.Copy(responseBuilder, readBuffer) // Copy initial response
+	}
+
+	reqBuilder := &strings.Builder{}
+
+	inputEvents := make(map[string]interface{})
+	for _, req := range r.Inputs {
+		reqBuilder.Grow(len(req.Data))
+
+		finalData, dataErr := expressions.EvaluateByte([]byte(req.Data), payloadValues)
+		if dataErr != nil {
+			r.options.Output.Request(r.options.TemplateID, input, "websocket", dataErr)
+			r.options.Progress.IncrementFailedRequestsBy(1)
+			return errors.Wrap(dataErr, "could not evaluate template expressions")
+		}
+		reqBuilder.WriteString(string(finalData))
+
+		err = wsutil.WriteClientMessage(conn, ws.OpText, finalData)
+		if err != nil {
+			r.options.Output.Request(r.options.TemplateID, input, "websocket", err)
+			r.options.Progress.IncrementFailedRequestsBy(1)
+			return errors.Wrap(err, "could not write request to server")
+		}
+
+		msg, _, err := wsutil.ReadServerData(conn)
```

Comment:
I wonder...how could we guarantee that the received message is indeed the response for our requests and not something else. We should probably also need to use the OpCode returned with the server message

Reply:
Added a validation which skips messages with non `text` or `binary` opcode.